plugins/renepay/flow.c

```diff
@@ -0,0 +1,849 @@
+		struct chan_extra *ce,
+		int dir,
+		struct amount_msat x)
+{
+	if(amount_msat_greater(x,ce->capacity))
+	{
+		debug_err("%s unexpected capacity=%s is less than x=%s",
+			__PRETTY_FUNCTION__,
+			type_to_string(tmpctx,struct amount_msat,&ce->capacity),
+			type_to_string(tmpctx,struct amount_msat,&x)
+			);
+		x = ce->capacity;
+	}
+
+	ce->half[dir].known_min = x;
+	ce->half[dir].known_max = x;
+
+	chan_extra_adjust_half(ce,!dir);
+}
+void chan_extra_set_liquidity(
+		struct chan_extra_map *chan_extra_map,
+		struct short_channel_id scid,
+		int dir,
+		struct amount_msat x)
+{
+	struct chan_extra *ce = chan_extra_map_get(chan_extra_map,
+						   scid);
+	if(!ce)
+	{
+		debug_err("%s unexpected chan_extra ce is NULL",
+			__PRETTY_FUNCTION__);
+	}
+	chan_extra_set_liquidity_(ce,dir,x);
+}
+/* Update the knowledge that this (channel,direction) has sent x msat.*/
+static void chan_extra_sent_success_(
+		struct chan_extra *ce,
+		int dir,
+		struct amount_msat x)
+{
+	if(amount_msat_greater(x,ce->capacity))
+	{
+		debug_err("%s unexpected capacity=%s is less than x=%s",
+			__PRETTY_FUNCTION__,
+			type_to_string(tmpctx,struct amount_msat,&ce->capacity),
+			type_to_string(tmpctx,struct amount_msat,&x)
+			);
+		x = ce->capacity;
+	}
+
+	struct amount_msat new_a, new_b;
+
+	if(!amount_msat_sub(&new_a,ce->half[dir].known_min,x))
+		new_a = AMOUNT_MSAT(0);
+	if(!amount_msat_sub(&new_b,ce->half[dir].known_max,x))
+		new_b = AMOUNT_MSAT(0);
+
+	ce->half[dir].known_min = new_a;
+	ce->half[dir].known_max = new_b;
+
+	chan_extra_adjust_half(ce,!dir);
+}
+void chan_extra_sent_success(
+		struct chan_extra_map *chan_extra_map,
+		struct short_channel_id scid,
+		int dir,
+		struct amount_msat x)
+{
+	struct chan_extra *ce = chan_extra_map_get(chan_extra_map,
+						   scid);
+	if(!ce)
+	{
+		debug_err("%s unexpected chan_extra ce is NULL",
+			__PRETTY_FUNCTION__);
+	}
+	chan_extra_sent_success_(ce,dir,x);
+}
+/* Forget a bit about this (channel,direction) state. */
+static void chan_extra_relax_(
+		struct chan_extra *ce,
+		int dir,
+		struct amount_msat down,
+		struct amount_msat up)
+{
+	struct amount_msat new_a, new_b;
+
+	if(!amount_msat_sub(&new_a,ce->half[dir].known_min,down))
+		new_a = AMOUNT_MSAT(0);
+	if(!amount_msat_add(&new_b,ce->half[dir].known_max,up))
+		new_b = ce->capacity;
+	new_b = amount_msat_min(new_b,ce->capacity);
+
+	ce->half[dir].known_min = new_a;
+	ce->half[dir].known_max = new_b;
+
+	chan_extra_adjust_half(ce,!dir);
+}
+void chan_extra_relax(
+		struct chan_extra_map *chan_extra_map,
+		struct short_channel_id scid,
+		int dir,
+		struct amount_msat x,
+		struct amount_msat y)
+{
+	struct chan_extra *ce = chan_extra_map_get(chan_extra_map,
+						   scid);
+	if(!ce)
+	{
+		debug_err("%s unexpected chan_extra ce is NULL",
+			__PRETTY_FUNCTION__);
+	}
+	chan_extra_relax_(ce,dir,x,y);
+}
+
+/* Forget the channel information by a fraction of the capacity. */
+void chan_extra_relax_fraction(
```

Comment:
I assume in practise fraction is a function depending on the time since we last learnt something about the channel?

Reply:
Yes. Right now it is just a linear function. Also you don't need to store the last time for each channel, because you can assume that each time you make a payment you improve knowledge about every channel even if that knowledge if zero for some. 

My intuition also tells me that whatever operation you choose, it's time dependence must correspond to a 1-dimensional semi-group:
```
Op(t1+t2) = Op(t1) * Op(t2)
```

Reply:
The one-parameter semi-group property of the operator is what allows you to ignore the "last time this channel was updated" and apply the operator on every channel every time there is a request for payment.